src/Netflix/Netflix.ts

```diff
@@ -0,0 +1,760 @@
+ * context
+ */
+
+enum NetflixPluginContextEnum {
+  watch = "Netflix Video Player Controls",
+  browse = "Browse Netflix"
+}
+type NetflixPluginContext = NetflixPluginContextEnum | null;
+
+const contextManager = (() => {
+  let enabled = true;
+
+  type ContextPatcher = {
+    add: (context: string) => void;
+    remove: (context: string) => void;
+  };
+
+  /**
+   * Patch context to avoid context duplication
+   * due to addContext being called multiple times causing duplicate
+   * context entry
+   *
+   * Usage:
+   *
+   * ```typescript
+   * patchContext((patcher) => {
+   *  patcher.add("ContextA")
+   *  patcher.add("ContextB")
+   *  patcher.add("ContextC")
+   * });
+   * ```
+   */
+  const patchContext = (fn: (patcher: ContextPatcher) => unknown) => {
+    const contextSet = new Set(PluginBase.util.getContext());
+    const contextPatcher = {
+      add: (context: string) => {
+        if (!contextSet.has(context)) {
+          PluginBase.util.addContext(context);
+        }
+      },
+      remove: (context: string) => {
+        if (contextSet.has(context)) {
+          PluginBase.util.removeContext(context);
+        }
+      }
+    } as ContextPatcher;
+    return fn(contextPatcher);
+  };
+
+  const createContextFromUrl = (url: URL): NetflixPluginContext => {
+    const { pathname } = url;
+    switch (true) {
+      case pathname.startsWith("/watch"):
+        return NetflixPluginContextEnum.watch;
+      case pathname.startsWith("/latest"):
+      case pathname.startsWith("/browse"):
+      case pathname.startsWith("/title"):
+        return NetflixPluginContextEnum.browse;
+      default:
+        return null;
+    }
+  };
+  const setContext = (context: NetflixPluginContext) =>
+    patchContext(patcher => {
+      switch (true) {
+        case context === NetflixPluginContextEnum.browse: {
+          patcher.add(NetflixPluginContextEnum.browse);
+          patcher.add("Normal");
+          return;
+        }
+        case context === NetflixPluginContextEnum.watch: {
+          patcher.add(NetflixPluginContextEnum.watch);
+          patcher.remove("Normal");
+          return;
+        }
+        default: {
+          patcher.remove(NetflixPluginContextEnum.watch);
+          patcher.remove(NetflixPluginContextEnum.browse);
+          patcher.add("Normal");
+          return;
+        }
+      }
+    });
+
+  const refreshCurrentContext = () => {
+    try {
+      setContext(createContextFromUrl(new URL(window.location.href)));
+    } catch (error) {
+      console.error(error);
+    }
+  };
+
+  return {
+    enable: async () => {
+      enabled = true;
+      while (enabled) {
+        refreshCurrentContext();
+        await new Promise(resolve => setTimeout(resolve, 1000));
+      }
+    },
+    disable: () => {
+      enabled = false;
+      PluginBase.util.enterContext(["Normal"]);
+    }
+  };
+})();
+
+// Netflix specific functionalities
+
+type Command =
```

Comment:
Hmm, what's the benefit of having these types? I would make arrays of groups of command names, then use those as subgroups in the `contexts` property. Ideally we would make the `contexts` property more typesafe by only allowing valid command names. Perhaps you can help me with that?

Reply:
This `Command` type is used both in the page context and the extension context. They act as discriminated unions to discriminate commands by its `tag`, which in this case is the field `key`. It's less about contexts and more about parsing commands. It helps to make sure I don't put the wrong Command in a function signature.

If you're familiar with source code translation, this is like the very basic principle of parsing, but done to javascript JSONObject, utilizing Typescript types, and without the `consume` recursion.

^ I actually get to learn that discriminated union and parsing are kinds of the same when reading the source-code of Svelte. It was an interesting read

> Perhaps you can help me with that?

Let's